docs_src/design/adr/application/0025-Record-and-Replay.md

```diff
@@ -0,0 +1,128 @@
+
+!!! note
+    Application Services do not receive data when the Functions Pipelines are stopped.
+
+### Record Endpoint
+
+This `POST` API will start recording data as specified in the [request Data Transfer Object (DTO)](#record-request-dto) defined below. The request handler will validate the DTO and then create a new Functions Pipeline and [Start the Functions Pipeline](https://docs.edgexfoundry.org/2.3/microservices/application/ApplicationServiceAPI/#makeitrun) to process incoming data. 
+
+The Functions Pipeline will contain the following pipeline functions in the following order
+
+- [Filter functions](https://docs.edgexfoundry.org/2.3/microservices/application/BuiltIn/#filtering) if filtering is needed, configured based on the DTO parameters.
+- [Batching pipeline function](https://docs.edgexfoundry.org/2.3/microservices/application/BuiltIn/#batching) configured based on the DTO parameters. This will be used to control the record duration/count. 
+- New pipeline function written to process the batched data once the batching threshold has been exceeded. This function will simply send the recorded data to an async function for processing. 
+
+The async  function receiving the data will first [stop the Functions Pipeline](https://docs.edgexfoundry.org/2.3/microservices/application/ApplicationServiceAPI/#makeitstop) and then save the data for later replay and/or export. It will also determine the list of unique Device Profile and Device Names from the data and store them along side the recorded data. Since app services can receive Events out of order per their timestamps, the saved Event data must be sorted by the Event timestamps. All data will saved in in-memory storage. 
+
+!!! note
+    Starting a new recording will overwrite any previous recorded data.
+
+#### Record Request DTO
+
+##### Duration
+
+Time duration in which to record data. Required if **Event Limit** is not specified.
+
+##### Event Limit 
+
+Maximum number `Events` to record. Required if **Duration** is not specified
+
+##### Include Device Profile Names 
+
+Optional list of Device Profile Names to filter for
+
+##### Include Device Names
+
+Optional list of Device Names to filter for
+
+##### Exclude Device Profile Names
+
+Optional list of Device Profile Names to filter out
+
+##### Exclude Device Names
+
+Optional list of Device Names to filter out
+
+### Replay endpoint
+
+This `POST` API will start replaying the recorded data as specified in the [request Data Transfer Object (DTO)](#replay-request-dto) defined below. The request handler will validate the DTO and that the appropriate Device Profiles and Devices from the data exist. It will then start an async Go function to handle the replay so the request doesn't timeout on long replays. 
```

Comment:
do we anticipate a replay to the end of the recording or could someone ask to stop the replay mid-stream?  If so, would assume we need APIs for that?

Reply:
Replay will do the complete recording. Canceling the replay is not a requirement in the approved UCR. As with above this could be a future enhancement base on user requests.